whitepaper.tex

```diff
@@ -765,115 +809,146 @@
-    define how a component can be invoked, or can invoke others. An
-    interface is either imported (invoking an interface of another
-    component) or exported (able to be invoked by another component's
-    imported interface), except for the shared-memory interface, which
-    is symmetric.
-  \item[Connectors] connect like interfaces by linking an importing with an
-    exporting interface. Connectors in CAmkES are always one-to-one,
-    but broadcast or multicast functionality can be implemented on top
-    of this model by building components that copy inputs to multiple outputs.
-  \end{description}
+  \autoref{f:microkit} shows these abstractions (yes, that's all!) The
+  \emph{protection domain} (PD) is the process abstraction: it
+  represents a program in execution in its own, isolated address
+  space. Two PDs can be connected by a \emph{communication channel}
+  and may also share \emph{memory regions} that are mapped in one or
+  more PDs' address spaces. A PD may be a VM (running a guest OS), but
+  other PDs are oblivious to that fact.
+
+  \begin{Chilli}
+    Microkit further simplifies programming by imposing an
+    event-handling model. Specifically, each PD has two or three entry
+    points: \code{init()}, \code{notified()} and, optionally,
+    \code{protected()}. These are provided by the programmer and invoked
+    by the framework on particular events: \code{init} is called exactly
+    once at system initialisation time and before any of the other
+    handlers are called. \code{notified()} is invoked if the PD at the
+    other end of the channel calls the \code{notify()} API on the
+    channel, and \code{protected()} is invoked if the PD at the other
+    end of a channel calls the \code{ppcall()} API.
+
+    PDs have scheduling parameters, which include a priority.
+    \code{notify()} calls are asynchronous: they never block, and the
+    when the receiver PD's \code{notified()} function is invoked depends
+    on priorities and whether the two PDs are executing on the same core
+    or not. These calls pass no data to the receiver other than the
+    identity of the notifier, and as such behave similarly to
+    interrupts. In contrast, a \code{ppcall()} behaves like a normal
+    function call, except that the code it executes happens to be in a
+    different PD. In particular, \code{ppcall()} is synchronous, i.e.\ it blocks
+    the caller until the callee returns from its \code{protected()}
+    handler.  It can also pass arguments and return a
+    result value. To prevent deadlock,  a \code{ppcall()} is only
+    allowed from lower to higher priority.
+
+    PDs providing a \code{protected()} handler are referred to as
+    server, and the PD invoking \code{ppcall()} is called a
+    client. Servers execute on the client's \emph{scheduling context}
+    (see \autoref{s:mcs}) and core, meaning that their usage of the CPU
+    resource is properly accounted to the client.
```

Comment:
This is only true in the case of a passive PD, which is currently an opt-in attribute.

Reply:
Yes, but shouldn't we encourage people to always use passive servers? I'm still looking for a convincing use case for active ones

Reply:
I guess I can add a Chilli paragraph...

Reply:
Adding a comment in parenthreses